trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/Boogie2ACSL.java

```diff
@@ -0,0 +1,657 @@
+		final String boogieId = expr.getIdentifier();
+		if (boogieId.equals(SFO.RES)) {
+			// TODO: Can we somehow get the return type here?
+			return new BacktranslatedExpression(new ACSLResultExpression());
+		} else if (mMapping.hasVar(boogieId, expr.getDeclarationInformation())) {
+			final Pair<String, CType> pair = mMapping.getVar(boogieId, expr.getDeclarationInformation());
+			if (isPresentInContext(pair.getFirst(), context)) {
+				final var range = getRangeForCType(pair.getSecond());
+				return new BacktranslatedExpression(new IdentifierExpression(pair.getFirst()), pair.getSecond(),
+						range.getFirst(), range.getSecond());
+			}
+		} else if (mMapping.hasInVar(boogieId, expr.getDeclarationInformation())) {
+			// invars can only occur in expressions as part of synthetic expressions, and then they represent oldvars
+			final Pair<String, CType> pair = mMapping.getInVar(boogieId, expr.getDeclarationInformation());
+			final var range = getRangeForCType(pair.getSecond());
+			return new BacktranslatedExpression(new OldValueExpression(new IdentifierExpression(pair.getFirst())),
+					pair.getSecond(), range.getFirst(), range.getSecond());
+		}
+		return null;
+	}
+
+	private static BacktranslatedExpression constructFloat(final BitvecLiteral sign, final BitvecLiteral exponent,
+			final BitvecLiteral fraction) {
+		// TODO: Should we rather represent this C-float using scientific notation (e.g. -1.57E13)?
+		final String bit = bitvecToString(sign) + bitvecToString(exponent) + bitvecToString(fraction);
+		final BigDecimal f = getDecimalFromBinaryString(bit);
+		return new BacktranslatedExpression(new RealLiteral(f.toPlainString()));
+	}
+
+	private static BigDecimal getDecimalFromBinaryString(final String binary) {
+		final int len = binary.length();
+		if (len == 32) {
+			final int intBits = Integer.parseUnsignedInt(binary, 2);
+			final float floatValue = Float.intBitsToFloat(intBits);
+			return BigDecimal.valueOf(floatValue);
+		} else if (len == 64) {
+			final long longBits = Long.parseUnsignedLong(binary, 2);
+			final double doubleValue = Double.longBitsToDouble(longBits);
+			return BigDecimal.valueOf(doubleValue);
+		} else {
+			throw new IllegalArgumentException("Unsupported length: " + len);
+		}
+	}
+
+	private static String bitvecToString(final BitvecLiteral lit) {
+		final String binStr = new BigInteger(lit.getValue()).toString(2);
+		assert binStr.length() <= lit.getLength() : "Binary string cannot be longer than bitvector literal length";
+		final int missingZeros = lit.getLength() - binStr.length();
+		if (missingZeros > 0) {
+			final String formatStr = "%" + lit.getLength() + "s";
+			return String.format(formatStr, binStr).replace(' ', '0');
+		}
+		return binStr;
+	}
+
+	// TODO: Currently we don't care about types here, since function applications should only occur in bitvector
+	// setting, where we should not need any casts.
```

Comment:
I did not check this in detail, but for functions like `bvadd` we do not require any casts for soundness, since the result of the function application is restricted by the number of bits, so we do a modulo instead. However, using ranges we might avoid unnecessary modulos.

Reply:
I am not too familiar with our bitvector translation, so I might discuss this with backtranslation with Matthias first.